File: src/patterns/creational/prototype/zara.nim
# internal
import
    patterns/creational/prototype/[
        clothing
    ]


type
    Zara* = ref object of Clothing


proc newZara*(brand: string, model: string, price: float, owner: string): Zara =
    return Zara(
        brand: brand,
        model: model,
        price: price,
        owner: owner
    )

proc copy*(self: Zara): Zara =
    return Zara(
        brand: self.getBrand(),
        model: self.getModel(),
        price: self.getPrice(),
        owner: self.getOwner()
    )

File: src/patterns/creational/prototype/zara.go
package prototype


type Zara struct {
	*Clothing
}


func NewZara(brand string, model string, price float32, owner string) *Zara {
	return &Zara{
		&Clothing{
			brand: brand,
			model: model,
			price: price,
			owner: owner,
		},
	}
}

func (self *Zara) Copy() *Zara {
	return &Zara{
		&Clothing{
			brand: self.GetBrand(),
			model: self.GetModel(),
			price: self.GetPrice(),
			owner: self.GetOwner(),
		},
	}
}